File: game/scripts/vscripts/heroes/meepo/geostrike.ts
LinkLuaModifier("modifier_meepo_geostrike_lua","heroes/meepo/geostrike.lua",LuaModifierType.LUA_MODIFIER_MOTION_NONE);
LinkLuaModifier("modifier_meepo_geostrike_debuff_lua","heroes/meepo/geostrike.lua",LuaModifierType.LUA_MODIFIER_MOTION_NONE);

class meepo_geostrike_lua extends CDOTA_Ability_Lua {
  GetIntrinsicModifierName() { return "modifier_meepo_geostrike_lua" }

}



class modifier_meepo_geostrike_lua extends CDOTA_Modifier_Lua   {
  IsHidden() {return true}
  IsPermanent() {return true}


  DeclareFunctions() {

    return [
      modifierfunction.MODIFIER_EVENT_ON_ATTACK_LANDED,
    ]
  }

  

  OnAttackLanded(keys) {

    if (keys.attacker == this.GetParent() && !keys.attacker.PassivesDisabled()) {
      let ability = this.GetAbility()
      let duration = ability.GetSpecialValueFor("duration_tooltip")
      let target = <CDOTA_BaseNPC> keys.target
      let attacker = <CDOTA_BaseNPC> keys.attacker
      let modifier = target.FindModifierByNameAndCaster("modifier_meepo_geostrike_debuff_lua",attacker)
      if (modifier) {
        modifier.SetDuration(duration,true)
      } else {
        target.AddNewModifier(attacker,this.GetAbility(),"modifier_meepo_geostrike_debuff_lua",{duration:duration})
      }
    }
  }
}

class modifier_meepo_geostrike_debuff_lua extends CDOTA_Modifier_Lua {
  damage : number
  slow : number
  particle:ParticleID

  GetAttributes() {
    return DOTAModifierAttribute_t.MODIFIER_ATTRIBUTE_MULTIPLE
    
  }

  DeclareFunctions() {
    return [
      modifierfunction.MODIFIER_PROPERTY_MOVESPEED_BONUS_PERCENTAGE,
    ]
  }


  GetModifierMoveSpeedBonus_Percentage() {
    return this.slow
  }
  OnCreated() {
    if (IsServer()) {
      this.damage = this.GetAbility().GetAbilityDamage()
      this.slow = this.GetAbility().GetSpecialValueFor("slow")
      this.StartIntervalThink(1)
    }
  }

  OnIntervalThink() {
    let dTable: DamageTable = {attacker:this.GetCaster(),victim:this.GetParent(),damage:this.damage,damage_type:DAMAGE_TYPES.DAMAGE_TYPE_MAGICAL};
    ApplyDamage(dTable);
  }

  GetEffectName() {
    return "particles/units/heroes/hero_meepo/meepo_geostrike.vpcf"
  }
  GetEffectAttachType() {
    return ParticleAttachment_t.PATTACH_ABSORIGIN_FOLLOW
  }
}
 

File: game/scripts/vscripts/heroes/meepo/geostrike.lua
LinkLuaModifier("modifier_meepo_geostrike_lua","heroes/meepo/geostrike.lua",LUA_MODIFIER_MOTION_NONE)
LinkLuaModifier("modifier_meepo_geostrike_debuff_lua","heroes/meepo/geostrike.lua",LUA_MODIFIER_MOTION_NONE)
meepo_geostrike_lua = class({})
function meepo_geostrike_lua.GetIntrinsicModifierName(self)
    return "modifier_meepo_geostrike_lua"
end
modifier_meepo_geostrike_lua = class({})
function modifier_meepo_geostrike_lua.IsHidden(self)
    return true
end
function modifier_meepo_geostrike_lua.IsPermanent(self)
    return true
end
function modifier_meepo_geostrike_lua.DeclareFunctions(self)
    return {MODIFIER_EVENT_ON_ATTACK_LANDED}
end
function modifier_meepo_geostrike_lua.OnAttackLanded(self,keys)
    if (keys.attacker==self.GetParent(self)) and not keys.attacker.PassivesDisabled(keys.attacker) then
        local ability = self.GetAbility(self)
        local duration = ability.GetSpecialValueFor(ability,"duration_tooltip")
        local target = keys.target
        local attacker = keys.attacker
        local modifier = target.FindModifierByNameAndCaster(target,"modifier_meepo_geostrike_debuff_lua",attacker)
        if modifier then
            modifier.SetDuration(modifier,duration,true)
        else
            target.AddNewModifier(target,attacker,self.GetAbility(self),"modifier_meepo_geostrike_debuff_lua",{["duration"]=duration})
        end
    end
end
modifier_meepo_geostrike_debuff_lua = class({})
function modifier_meepo_geostrike_debuff_lua.GetAttributes(self)
    return MODIFIER_ATTRIBUTE_MULTIPLE
end
function modifier_meepo_geostrike_debuff_lua.DeclareFunctions(self)
    return {MODIFIER_PROPERTY_MOVESPEED_BONUS_PERCENTAGE}
end
function modifier_meepo_geostrike_debuff_lua.GetModifierMoveSpeedBonus_Percentage(self)
    return self.slow
end
function modifier_meepo_geostrike_debuff_lua.OnCreated(self)
    if IsServer() then
        self.damage=self.GetAbility(self).GetAbilityDamage(self.GetAbility(self))
        self.slow=self.GetAbility(self).GetSpecialValueFor(self.GetAbility(self),"slow")
        self.StartIntervalThink(self,1)
        self.particle=ParticleManager.CreateParticle(ParticleManager,"particles/units/heroes/hero_meepo/meepo_geostrike.vpcf",PATTACH_ABSORIGIN_FOLLOW,self.GetCaster(self))
        ParticleManager.SetParticleControl(ParticleManager,self.particle,0,Vector(0,0,0))
        self.AddParticle(self,self.particle,true,false,1,false,false)
    end
end
function modifier_meepo_geostrike_debuff_lua.OnIntervalThink(self)
    local dTable = {["attacker"]=self.GetCaster(self),["victim"]=self.GetParent(self),["damage"]=self.damage,["damage_type"]=DAMAGE_TYPE_MAGICAL}
    ApplyDamage(dTable)
end
